cmd/deploy.go

```diff
@@ -185,59 +201,133 @@ func runDeploy(cmd *cobra.Command, _ []string, newClient ClientFactory) (err err
 		err = errors.New("unrecognized builder: valid values are: s2i, pack")
 		return
 	}
-
-	// Use the user-provided builder image, if supplied
 	if config.BuilderImage != "" {
-		function.BuilderImages[config.Builder] = config.BuilderImage
+		f.BuilderImages[config.Builder] = config.BuilderImage
 	}
-
-	client, done := newClient(ClientConfig{Namespace: config.Namespace, Verbose: config.Verbose},
+	client, done := newClient(ClientConfig{Namespace: f.Namespace, Verbose: config.Verbose},
 		fn.WithRegistry(config.Registry),
 		fn.WithBuilder(builder))
 	defer done()
 
-	switch currentBuildType {
-	case fn.BuildTypeLocal, "":
-		if config.GitURL != "" || config.GitDir != "" || config.GitBranch != "" {
-			return fmt.Errorf("remote git arguments require the --build=git flag")
-		}
-		if err := client.Build(cmd.Context(), config.Path); err != nil {
-			return err
-		}
-	case fn.BuildTypeGit:
-		git := function.Git
+	// Default Client Registry, Function Registry or explicit Image required
+	//
+	// An attempt to build without either an image or registry
+	// will fail with an error.  For interactive prompting, use the -c flag.
+	// Since the error message generated by the client library in this case is a
+	// bit terse, let's check here and offer the user a very descriptive message
+	// which includes cli-specific recommendations
+	if client.Registry() == "" && f.Registry == "" && f.Image == "" {
+		return ErrRegistryRequired
+	}
 
+	// NOTE: curently need to preemptively write out Function state until
+	// the API is updated to use instances.
+	//
+	// Discussion:  The need for this is proof that the Client API should work on
+	// Function instances as opposed to paths.  In order for the following Build
+	// call to complete using flag values from above, the Function hast to be
+	// first serialized out to disk.  This interferes with the ability to have a
+	// --save option which only writes mutated values to disk on a successful
+	// completion.. i.e. we have to save to communicate the Functin state to the
+	// builder, potentially prematurely.  Therefore, in the forthcoming PR which
+	// persists all flags to func.yaml by default (option to --save=false), the
+	// Client API will need to change to accepting Function instances rather than
+	// paths on disk.  This has other knock-on benefits as well.
+	if err = f.Write(); err != nil {
+		return
+	}
```

Comment:
@zroubalik I believe this directly pertains to your question in the general discussion thread.

Reply:
Yeah, I agree with the proposal for the Client API to work on Function instances, no matter if we go with the `--save` flag apporach.